memiavl/snapshot.go

```diff
@@ -300,92 +338,84 @@ func (t *Tree) WriteSnapshot(snapshotDir string) (returnErr error) {
-	nodeIndex                             uint32
-	keysOffset, valuesOffset              uint64
+
+	// record the current node index and leaf node index
+	nodeIndex, leafIndex uint32
+
+	// record the current writing offset in keys/values file
+	keysOffset, valuesOffset uint64
+
+	// record the offsets for keys and values
+	keysOffsetBM, valuesOffsetBM *roaring64.Bitmap
 }
 
 func newSnapshotWriter(nodesWriter, keysWriter, valuesWriter io.Writer) *snapshotWriter {
 	return &snapshotWriter{
-		nodesWriter:  nodesWriter,
-		keysWriter:   keysWriter,
-		valuesWriter: valuesWriter,
+		nodesWriter:    nodesWriter,
+		keysWriter:     keysWriter,
+		valuesWriter:   valuesWriter,
+		keysOffsetBM:   roaring64.New(),
+		valuesOffsetBM: roaring64.New(),
 	}
 }
 
-// writeKey append key to keys file
-func (w *snapshotWriter) writeKey(key []byte) (uint64, error) {
-	var buf [SizeKeyLen]byte
-	if len(key) > math.MaxUint32 {
-		return 0, fmt.Errorf("key length overflow: %d", len(key))
-	}
-	binary.LittleEndian.PutUint32(buf[:], uint32(len(key)))
-	if _, err := w.keysWriter.Write(buf[:]); err != nil {
-		return 0, err
-	}
+// writeKey append key to keys file and record the offset
+func (w *snapshotWriter) writeKey(key []byte) error {
 	if _, err := w.keysWriter.Write(key); err != nil {
-		return 0, err
+		return err
 	}
-	offset := w.keysOffset
-	w.keysOffset += SizeKeyLen + uint64(len(key))
-	return offset, nil
+	w.keysOffsetBM.Add(w.keysOffset)
+	w.keysOffset += uint64(len(key))
+	return nil
 }
 
-// writeValue append value to values file
-func (w *snapshotWriter) writeValue(value []byte) (uint64, error) {
-	var buf [SizeValueLen]byte
-	if len(value) > math.MaxUint32 {
-		return 0, fmt.Errorf("value length overflow: %d", len(value))
-	}
-	binary.LittleEndian.PutUint32(buf[:], uint32(len(value)))
-	if _, err := w.valuesWriter.Write(buf[:]); err != nil {
-		return 0, err
-	}
+// writeValue append value to values file and record the offset
+func (w *snapshotWriter) writeValue(value []byte) error {
 	if _, err := w.valuesWriter.Write(value); err != nil {
-		return 0, err
+		return err
 	}
-	offset := w.valuesOffset
-	w.valuesOffset += SizeValueLen + uint64(len(value))
-	return offset, nil
+	w.valuesOffsetBM.Add(w.valuesOffset)
+	w.valuesOffset += uint64(len(value))
+	return nil
 }
 
 // writeRecursive write the node recursively in depth-first post-order,
 // returns `(nodeIndex, offset of minimal key in subtree, err)`.
-func (w *snapshotWriter) writeRecursive(node Node) (uint32, uint64, error) {
+func (w *snapshotWriter) writeRecursive(node Node) (uint32, uint32, error) {
 	var (
-		buf              [SizeNodeWithoutHash]byte
-		minimalKeyOffset uint64
+		buf [SizeNodeWithoutHash]byte
+		// record the minimal key index in the subtree, used to update key field in parent node.
+		minimalKeyIndex uint32
 	)
 
 	buf[OffsetHeight] = byte(node.Height())
 	if node.Version() > math.MaxUint32 {
 		return 0, 0, fmt.Errorf("version overflow: %d", node.Version())
 	}
```

Comment:
seems no need check after change Version to `uint32`
```suggestion
```

Reply:
done